server/util/rexec/rexec.go

```diff
@@ -0,0 +1,452 @@
+	client repb.ExecutionClient
+	stream repb.Execution_ExecuteClient
+	name   string
+}
+
+func NewRetryingStream(ctx context.Context, client repb.ExecutionClient, stream repb.Execution_ExecuteClient, name string) *RetryingStream {
+	ctx, cancel := context.WithCancel(ctx)
+	return &RetryingStream{
+		ctx:    ctx,
+		cancel: cancel,
+		client: client,
+		stream: stream,
+		name:   name,
+	}
+}
+
+// Name returns the operation name, if known.
+func (s *RetryingStream) Name() string {
+	return s.name
+}
+
+// Recv attempts to reliably return the next operation on the named stream.
+//
+// If the stream is disconnected and the operation name has been received, it
+// will attempt to reconnect with WaitExecution.
+func (s *RetryingStream) Recv() (*longrunning.Operation, error) {
+	r := retry.DefaultWithContext(s.ctx)
+	for {
+		op, err := s.stream.Recv()
+		if err == nil {
+			if op.GetName() != "" {
+				s.name = op.GetName()
+			}
+			return op, nil
+		}
+		if !status.IsUnavailableError(err) || s.name == "" {
+			return nil, err
+		}
+		if !r.Next() {
+			return nil, s.ctx.Err()
+		}
+		req := &repb.WaitExecutionRequest{Name: s.name}
+		next, err := s.client.WaitExecution(s.ctx, req)
+		if err != nil {
+			return nil, err
+		}
+		s.stream.CloseSend()
+		s.stream = next
+	}
+}
+
+func (s *RetryingStream) CloseSend() error {
+	var err error
+	if s.stream != nil {
+		err = s.stream.CloseSend()
+		s.stream = nil
+	}
+	s.client = nil
+	s.cancel()
+	return err
+}
+
+// ExecuteOperation contains an operation along with its execution-specific
+// payload.
+type ExecuteOperation struct {
+	*longrunning.Operation
+
+	// ExecuteOperationMetadata contains any metadata unpacked from the
+	// operation.
+	ExecuteOperationMetadata *repb.ExecuteOperationMetadata
+	// ExecuteResponse contains any response unpacked from the operation.
+	ExecuteResponse *repb.ExecuteResponse
+	// Err contains any error parsed from the ExecuteResponse status field.
+	Err error
+}
+
+// UnpackOperation unmarshals all expected execution-specific fields from the
+// given operationn.
+func UnpackOperation(op *longrunning.Operation) (*ExecuteOperation, error) {
+	msg := &ExecuteOperation{Operation: op}
+	if op.GetResponse() != nil {
+		msg.ExecuteResponse = &repb.ExecuteResponse{}
+		if err := op.GetResponse().UnmarshalTo(msg.ExecuteResponse); err != nil {
+			return nil, err
+		}
+	}
+	if op.GetMetadata() != nil {
+		msg.ExecuteOperationMetadata = &repb.ExecuteOperationMetadata{}
+		if err := op.GetMetadata().UnmarshalTo(msg.ExecuteOperationMetadata); err != nil {
+			return nil, err
+		}
+	}
+	msg.Err = gstatus.FromProto(msg.ExecuteResponse.GetStatus()).Err()
+	return msg, nil
+}
+
+// AssemblePlatform assembles a Platform proto from a list of NAME=VALUE pairs.
+// If the same name is specified more than once, the last one wins. The entries
+// are sorted by name, so that the platform is cache-friendly.
+func AssemblePlatform(pairs ...string) (*repb.Platform, error) {
```

Comment:
should this be private? I searched the other CL to see where it was used but did not find it.

Reply:
In the latest refactor, clients will use this to assemble the Command proto.